archive/imo/imo1969_q1.lean

```diff
@@ -17,37 +17,36 @@ open nat
 Prove that there are infinitely many natural numbers $a$ with the following property:
 the number $z = n^4 + a$ is not prime for any natural number $n$.
 
-The key to the solution is that you can factor z into the product of two polynomials,
-if a = 4*m^4.
+The key to the solution is that you can factor $z$ into the product of two polynomials,
+if $a = 4*m^4$. This is Sophie Germain's identity, called `pow_four_add_four_mul_pow_four`
+in mathlib.
 -/
-
-lemma factorization (m n : ℤ) : ((m - n)^2 + m^2) * ((m + n)^2 + m^2) = n^4 + 4*m^4 := by ring
+lemma factorization {m n : ℤ} : ((n - m)^2 + m^2) * ((n + m)^2 + m^2) = n^4 + 4*m^4 :=
+pow_four_add_four_mul_pow_four.symm
 
 /-!
 To show that the product is not prime, we need to show each of the factors is at least 2,
-which nlinarith can solve since they are expressed as a sum of squares.
+which `nlinarith` can solve since they are expressed as a sum of squares.
 -/
-lemma left_factor_large (m n : ℤ) (h: 1 < m) : 1 < ((m - n)^2 + m^2) := by nlinarith
-lemma right_factor_large (m n : ℤ) (h: 1 < m) : 1 < ((m + n)^2 + m^2) := by nlinarith
+lemma left_factor_large {m : ℤ} (n : ℤ) (h : 1 < m) : 1 < ((n - m)^2 + m^2) := by nlinarith
+lemma right_factor_large {m : ℤ} (n : ℤ) (h : 1 < m) : 1 < ((n + m)^2 + m^2) := by nlinarith
```

Comment:
What's the difference between the types of braces here?

Reply:
Curly braces surround arguments that are implicit, which we generally use when Lean can infer them from later arguments, see: https://leanprover.github.io/theorem_proving_in_lean/dependent_type_theory.html#implicit-arguments